### PR TITLE
Js group project

### PR DESCRIPTION
**Pull request by AlyAtalla**
Use JavaScript to make websites dynamic and build basic single page apps.
Use ES6 syntax.
Use ES6 modules.
Use callbacks and promises.
Use webpack.
Apply JavaScript best practices and language style guides in code.
Use AAA pattern for unit tests.
Write units tests for a JavaScript app.
Follow Gitflow.
Solve simple git conflicts.
Send and receive data from an API.
Use API documentation.
Understand and use JSON.
Make JavaScript code asynchronous.

Presentation video:
https://youtu.be/u-O1FAvcjR8

### CODE REVIEW

File: __tests__/items-count.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+const itemsCounter = () => itemsContainer.getElementsByTagName('item-card').length;
+
+describe('items-counter-tests', () => {
+  it('should be equal to fetched items length', () => {
+    const items = getItems();
+    displayItems();
+
+    expect(itemsCounter()).toBe(items.length);
+  });
+});
```
Comment (nahnahdev) [inline __tests__/items-count.js]: - [ ] Great job so far @AlyAtalla and @NalianAchei2020 You have done a very impressive job so far. 
- [ ] Your current tests checks if the `counter function` returns the expected count when items are displayed. However, it would be helpful to add tests for edge cases where there are no items displayed. :+1:  
- [x] You can create a test that checks if the `counter function` returns `0 ` when there are no items in the container :wink:

File: __tests__/comment-count.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    commentCounter();
+    expect(comments.querySelector('#comments-count').textContent).toBe('(2)');
+  });
+  it('should return 3', () => {
+    postComment();
+    appendComments(getComments());
+    commentCounter();
+    expect(comments.querySelector('#comments-count').textContent).toBe('(3)');
+  });
+});
```
Comment (nahnahdev) [inline __tests__/comment-count.js]: - [ ] Just as I mentioned earlier, I noticed here that your test cases cover the basic functionality of the `commentCounter` function. However, it will be good to add a test for the scenario where no comments are displayed initially (i.e., empty comments array) :wink: 
- [ ] Additionally, you can test the scenario when a comment is posted and the comment count is updated. :+1:

File: __tests__/items-count.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+const document = new JSDOM().window.document;
+
+const itemsContainer = document.createElement('div');
+document.body.appendChild(itemsContainer);
+
+// mock get items
+const getItems = () => Array(24).fill(0);
+
+const displayItems = () => {
```
Comment (nahnahdev) [inline __tests__/items-count.js:11]: - [ ] Currently, your tests are failing. So kindly fix that as well :wink: 

image

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @AlyAtalla and @NalianAchei2020 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- Linters well setup and passing :heavy_check_mark: 
- Good use of GitFlow :heavy_check_mark: 
- Most required tests included :heavy_check_mark: 
- Comments feature and likes feature work as  expected :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: __tests__/items-count.js (hunk 3 above)
Comment (AlyAtalla) [reply __tests__/items-count.js:11]: Thank you.

File: __tests__/comment-count.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    expect(comments.querySelector('#comments-count').textContent).toBe('(2)');
+  });
+
+  it('should return 3', () => {
+    postComment();
+    appendComments(getComments());
+    commentCounter();
+    expect(comments.querySelector('#comments-count').textContent).toBe('(3)');
+  });
+});
```
Comment (Gambit142) [inline __tests__/comment-count.js]: - [ ] As highlighted by the previous reviewer, kindly write a test that shows the comment counter displays **_zero(0)_** when there are no comments for an item. By writing this test, you will be testing for edge cases. By writing test cases that cover different scenarios and edge cases, developers can detect and fix bugs before they impact the application in production. 👍 😎

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @AlyAtalla @NalianAchei2020 👋, 

Great work on making the changes requested by a previous reviewer 👏 

## Highlights
- All linters checks passed ✔️ 
- Application is identical to the wireframe provided ✔️ 
- Used Jest testing library ✔️ 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by OyePriscilla — APPROVED** [review]
Hi @AlyAtalla and @NalianAchei2020 👏 👏, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
Added test case for null comment ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
